File: 112-PathSum/112-PathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def dfs(root, targetSum, currSum):
            if root == None: return False

            currSum = currSum + root.val

            if root.left == None and root.right == None:
                if currSum == targetSum: return True 

            return dfs(root.left, targetSum, currSum) or dfs(root.right, targetSum, currSum)

        
        return dfs(root, targetSum, 0)

File: 112-PathSum/112-PathSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    if (root == null) return false;

    stack = [[root, 0]];

    while(stack.length) {
        let [node, acc] = stack.pop()
         
        acc += node.val

        if(!node.left && !node.right){
            if(acc == targetSum) return true
        }

        if(node.left) stack.push([node.left, acc])
        if(node.right) stack.push([node.right, acc])
    }

    return false;
};